### PR TITLE
Fix callable entry points

### LINKED ISSUE #2564: Problems with hypothesis entrypoints
Currently I am testing out this new cool `entrypoints` feature. #2556 
And it does not work correctly for me.
I specify an entrypoint and nothing happens. The entrypoint function is not getting called.

I am following this example: https://github.com/HypothesisWorks/hypothesis/commit/4ca04cf85b1389f0233dc652a5a74f15d2922c05#diff-94fd9cebb412c332ed91ff584ee7c2f3R129-R155

The problem is best shown by this piece of code:

```python
# module.py

print('Module imported!')

def _hypothesis_setup_hook():
    print('func called')
```

Then I add it to the `setup.py` the same way it is shown in the docs.
Now, let's try to call `run()` function:

```python
>>> for entry_point in pkg_resources.iter_entry_points("hypothesis"):
...     print(entry_point)
...     entry_point.load()
```

Output:

```
_ = mymodule:_hypothesis_setup_hook
Module imported!
<function _hypothesis_setup_hook at 0x100f20ca0>
```

When I change this code to actually execute a function, it works:

```python
>>> for entry_point in pkg_resources.iter_entry_points("hypothesis"):
...     entry_point.load()()  # calling it now
```

Outputs:

```
Module imported!
func called
```

### LINKED ISSUE #2571: Avoid importing pkg_resources
Hypothesis recently started importing `pkg_resources` which is [notoriously slow](https://github.com/pypa/setuptools/issues/510) to import. A better alternative these days is to use [`importlib.metadata`](https://docs.python.org/3/library/importlib.metadata.html), added in Python 3.8, or its backport `importlib_metadata`. Doing so would look something like this (completely untested, just to get the idea across):

```diff
diff --git a/hypothesis-python/src/hypothesis/entry_points.py b/hypothesis-python/src/hypothesis/entry_points.py
index 8cc4137ba..a5d670143 100644
--- a/hypothesis-python/src/hypothesis/entry_points.py
+++ b/hypothesis-python/src/hypothesis/entry_points.py
@@ -20,9 +20,18 @@ custom types, running the relevant code when *hypothesis* is imported instead of
 your package.
 """
 
-import pkg_resources
+try:
+    if sys.version_info >= (3, 8):
+        from importlib import metadata as importlib_metadata
+    else:
+        import importlib_metadata
+except ImportError:
+    import pkg_resources
 
-
-def run():
-    for entry_point in pkg_resources.iter_entry_points("hypothesis"):
-        entry_point.load()  # pragma: no cover
+    def run():
+        for entry_point in pkg_resources.iter_entry_points("hypothesis"):
+            entry_point.load()  # pragma: no cover
+else:
+    def run():
+        for entry_point in importlib_metadata.entry_points().get("hypothesis", []):
+            entry_point.load()  # pragma: no cover
```

### PR DESCRIPTION
**Pull request by Zac-HD**
Fixes #2564, where apparently all my tests only covered import-time side effects :grimacing: 

Also fixes #2571 by trying various ways to get entry points from most to least performant.

### CODE REVIEW

**Review by sobolevn — APPROVED** [review]

**Comment by DRMacIver** [comment]
Hey @Zac-HD, remember when I said "We can always add a test when it breaks?"... 😉

**Review by sobolevn — COMMENTED** [review]
There's one more thing.

Example provided in the docs: https://hypothesis.readthedocs.io/en/latest/strategies.html?highlight=setuptools#registering-strategies-via-setuptools-entry-points
Actually breaks coverage.

I had this code:

```python
from returns.result import Result

def hypothesis_setup_hook():
    from hypothesis import strategies as st

    st.register_type_strategy(Result, st.one_of(
        st.builds(Result.from_value),
        st.builds(Result.from_failure),
    ))
```

And all things imported with `Result` was not covered.

> returns/result.py                              120     76      8      0    38%

Moving `import` inside the function does not help either.

Or should it be a separate issue?

**Comment by Zac-HD** [comment]
@sobolevn - my first guess is that coverage measurement is starting after the entry points are set up.  You could just specify the module form, though, and see if it works that way.

**Comment by sobolevn** [comment]
>  You could just specify the module form, though, and see if it works that way

As I understood it:

```python
from returns import result

def hypothesis_setup_hook():
    from hypothesis import strategies as st

    st.register_type_strategy(result.Result, st.one_of(
        st.builds(result.Result.from_value),
        st.builds(result.Result.from_failure),
    ))
```

Nope, coverage still does not work correctly. Instead of `100%` coverage `38%` is reported.

**Comment by sobolevn** [comment]
For any other person trying to fix `coverage`. Here's the solution:

```python
import sys
from hypothesis import strategies as st

from returns.contrib.hypothesis.containers import create_container_factory

def _setup():
    from returns.result import Result  # coverage for this module is broken
    st.register_type_strategy(Result, create_container_factory(Result))  # just our own custom strategy

    to_clean = []  # cleaning up our all imported modules, except the entrypoint itself: 
    for module in sys.modules:
        if module.startswith('returns.') and not module.startswith('returns.contrib.hypothesis'):
            to_clean.append(module)

    for module in to_clean:
        sys.modules.pop(module, None)  # cleaning up modules allows coverage to re-import them and cover properly

_setup()
del _setup
```

https://github.com/dry-python/returns/blob/master/returns/contrib/hypothesis/_entrypoint.py

**Comment by sobolevn** [comment]
I have made a simple reproduction:
- Source: https://github.com/sobolevn/hypothesis-coverage-bug
- Demo output with `0%` coverage: https://github.com/sobolevn/hypothesis-coverage-bug/runs/1104313120?check_suite_focus=true#step:5:31

**Comment by sobolevn** [comment]
Discussion: https://github.com/sobolevn/hypothesis-coverage-bug/issues/1

**Comment by sobolevn** [comment]
Yeap, the problem is in `pytest` plugins ordering. 
Temporary solution: https://github.com/sobolevn/hypothesis-coverage-bug/issues/1#issuecomment-691448550

Related https://github.com/pytest-dev/pytest/issues/935

File: hypothesis-python/docs/strategies.rst [hunk 1]
```diff
@@ -152,8 +152,27 @@ and then tell ``setuptools`` that this is your ``"hypothesis"`` entry point:
...
+    entry_points = {"hypothesis": ["_ = mymodule.a_submodule"]}
+
+    # Or name a specific function too, and Hypothesis will call it for you
     entry_points = {"hypothesis": ["_ = mymodule:_hypothesis_setup_hook"]}
-    ...
 
 And that's all it takes!
+
+
+Interaction with :pypi:`pytest-cov`
```
Comment (sobolevn) [inline hypothesis-python/docs/strategies.rst:164]: 👍

**Review by sobolevn — APPROVED** [review]
Looks like tests are gone! 🐑

**Comment by Zac-HD** [comment]
OK.  So far as I can tell, this is working... but I can't get automated tests to work :sob: 

https://github.com/Zac-HD/hypothesis/commit/e296edb51bdfafbc586287cc4ab86d069be87fbf got close-ish, but I think we'd have to actually ship and install multiple micro-packages just for testing and that would not be fun.

File: hypothesis-python/docs/strategies.rst (hunk 1 above)
Comment (Zac-HD) [reply hypothesis-python/docs/strategies.rst:164]: Can you cherry-pick and try out https://github.com/Zac-HD/hypothesis/commit/3503e3a4e4f9d0af45812de58b1eef9c582b7145?  The idea is that deferring imports for long enough in our pytest plugin would mean that Hypothesis entry points are only loaded after pytest has loaded all their entrypoints, but if it doesn't work for your workload it's *definitely* not worth the trouble.

**Comment by sobolevn** [comment]
Looks like the easiest way is to create three files: 
- `examples/setup.py` with `entrypoints` definition
- `examples/entrypoint.py` as an actual entrypoint
- `examples/tests/test_entrypoint.py` to test that entrypoint worked. Like I do here: https://github.com/dry-python/returns/blob/master/tests/test_contrib/test_hypothesis/test_type_resolution.py I just test that types declared in `entrypoint.py` do resolve

And then run this test in the CI separately.

**Comment by Zac-HD** [comment]
:tada: 

Thanks again to @sobolevn for all your feedback and help with this one :heart:

**Comment by sobolevn** [comment]
Thanks a lot for solving it! 👍

**Pull request merged**